Add support for the “details” element’s “name” attribute

sideshowbarker opened the pull request:
#### d4b4fa5cfed7b41046a9e62017e7a6e787d9d8ba
<pre>
Add support for the “details” element’s “name” attribute
<a href="https://bugs.webkit.org/show_bug.cgi?id=260893">https://bugs.webkit.org/show_bug.cgi?id=260893</a>

Reviewed by Tim Nguyen.

This changes adds support for the “name” attribute for the “details”
element, which allows grouping together a set of “details” elements such
that opening one “details” element in a group causes all other “details”
elements in the group to close.

<a href="https://github.com/whatwg/html/pull/9400">https://github.com/whatwg/html/pull/9400</a>

This change also introduces the DetailsNameAttributeEnabled preference
(enabled by default), for controlling whether the “name” attribute
actually causes the grouping/closing behavior described above.

* LayoutTests/imported/w3c/resources/resource-files.json:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-color-mix-function-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/elements/name-content-attribute-and-property-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/elements/name-content-attribute-and-property.html:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/name-attribute.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/name-attribute.tentative.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/support/empty-html-document.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/support/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/w3c-import.log:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-color-mix-function-expected.txt: Renamed from LayoutTests/platform/mac-monterey/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-color-mix-function-expected.txt.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/html/HTMLDetailsElement.cpp:
(WebCore::HTMLDetailsElement::attributeChanged):
* Source/WebCore/html/HTMLDetailsElement.idl:

Canonical link: <a href="https://commits.webkit.org/267756@main">https://commits.webkit.org/267756@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1f3b03a13802b8f4257e6c93bbb2c83952b2de64

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17614 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17939 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18469 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19423 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16472 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17807 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21225 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18085 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18569 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17825 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18126 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/15290 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20267 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/15361 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16035 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/22634 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/15220 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16370 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16204 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20490 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/16820 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16778 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14209 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/19186 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/15896 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/15876 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/4454 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4192 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/20262 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/20419 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16621 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/4295 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -159,6 +159,15 @@ void HTMLDetailsElement::attributeChanged(const QualifiedName& name, const AtomS
                 root->appendChild(*m_defaultSlot);
                 queueDetailsToggleEventTask(DetailsState::Closed, DetailsState::Open);
+                if (auto& name = attributeWithoutSynchronization(nameAttr); !name.isEmpty()) {
+                    Vector<HTMLDetailsElement*> otherDetailsElementsInThisNameGroup;
```
```suggestion
                    Vector<RefPtr<HTMLDetailsElement>> otherElementsInNameGroup;
```

nt1m commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -159,6 +159,15 @@ void HTMLDetailsElement::attributeChanged(const QualifiedName& name, const AtomS
+                        if (&detailsElement != this && detailsElement.attributeWithoutSynchronization(nameAttr) == name)
+                            otherDetailsElementsInThisNameGroup.append(&detailsElement);
+                    }
+                    for (HTMLDetailsElement* otherDetailsElement : otherDetailsElementsInThisNameGroup)
```
```suggestion
                    for (RefPtr otherDetailsElement : otherDetailsElementsInThisNameGroup)
```

nt1m commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -159,6 +159,15 @@ void HTMLDetailsElement::attributeChanged(const QualifiedName& name, const AtomS
             if (m_isOpen) {
                 root->appendChild(*m_defaultSlot);
                 queueDetailsToggleEventTask(DetailsState::Closed, DetailsState::Open);
+                if (auto& name = attributeWithoutSynchronization(nameAttr); !name.isEmpty()) {
```
Can we gate this code behind a preference (enabled by default is fine, I just want to make sure we have an off switch for the feature if needed)? See https://searchfox.org/wubkat/search?q=inertAttributeEnabled&path=Source&case=false&regexp=false for instance.

nt1m commented on Source/WebCore/html/HTMLDetailsElement.idl:
```diff
@@ -20,6 +20,7 @@
 [
     Exposed=Window
 ] interface HTMLDetailsElement : HTMLElement {
+    [CEReactions=NotNeeded, Reflect] attribute DOMString name;
```
You can use `EnabledBySetting` to gate this behind a preference.

nt1m commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -2205,6 +2205,20 @@ DeprecationReportingEnabled:
+    WebKit:
+      default: true
+    WebCore:
+      default: true
```
Upon reflection, can we enable it only in STP ? Mostly to assess the state of the spec before enabling in stable releases.

```suggestion
DetailsNameAttributeEnabled:
  type: bool
  status: preview
  category: html
  humanReadableName: "HTML <details name> attribute"
  humanReadableDescription: "Enable HTML <details name> attribute"
  defaultValue:
    WebKitLegacy:
      default: false
    WebKit:
      default: false
    WebCore:
      default: false
```

nt1m reviewed (commented):
Looks good, can you add a link to the HTML spec PR in the commit message?

sideshowbarker committed:
Add support for the “details” element’s “name” attribute
https://bugs.webkit.org/show_bug.cgi?id=260893

Reviewed by Tim Nguyen.

This changes adds support for the “name” attribute for the “details”
element, which allows grouping together a set of “details” elements such
that opening one “details” element in a group causes all other “details”
elements in the group to close.

https://github.com/whatwg/html/pull/9400

This change also introduces the DetailsNameAttributeEnabled preference
(enabled by default), for controlling whether the “name” attribute
actually causes the grouping/closing behavior described above.

* LayoutTests/imported/w3c/resources/resource-files.json:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-color-mix-function-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/elements/name-content-attribute-and-property-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/elements/name-content-attribute-and-property.html:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/name-attribute.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/name-attribute.tentative.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/support/empty-html-document.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/support/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/w3c-import.log:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-color-mix-function-expected.txt: Renamed from LayoutTests/platform/mac-monterey/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-color-mix-function-expected.txt.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/html/HTMLDetailsElement.cpp:
(WebCore::HTMLDetailsElement::attributeChanged):
* Source/WebCore/html/HTMLDetailsElement.idl:

Canonical link: https://commits.webkit.org/267756@main
